```
$ nix-shell -p neofetch --run neofetch --argstr system aarch64-darwin                                                                       
                    'c.          angerman@aarch64-darwin-1
                 ,xNMM.          -------------------------
               .OMMMMo           OS: macOS 12.1 21C52 arm64
               OMMM0,            Host: Macmini9,1
     .;loddo:' loolloddol;.      Kernel: 21.2.0
   cKMMMMMMMMMMNWMMMMMMMMMM0:    Uptime: 16 hours, 19 mins
 .KMMMMMMMMMMMMMMMMMMMMMMMWd.    Packages: 176 (nix-system)
 XMMMMMMMMMMMMMMMMMMMMMMMX.      Shell: bash 5.1.8
;MMMMMMMMMMMMMMMMMMMMMMMM:       DE: Aqua
:MMMMMMMMMMMMMMMMMMMMMMMM:       WM: Quartz Compositor
.MMMMMMMMMMMMMMMMMMMMMMMMX.      WM Theme: Blue (Light)
 kMMMMMMMMMMMMMMMMMMMMMMMMWd.    Terminal: /dev/ttys001
 .XMMMMMMMMMMMMMMMMMMMMMMMMMMk   CPU: Apple M1
  .XMMMMMMMMMMMMMMMMMMMMMMMMK.   GPU: Apple G13G
    kMMMMMMMMMMMMMMMMMMMMMMd     Memory: 1434MiB / 16384MiB
     ;KMMMMMMMWXXWMMMMMMMk.
       .cooc,.    .,coo:.
```
Getting a cached build from hydra works (well, "works", it pulls the binary from cache...)
```
$ nix-build "<nixpkgs>" -A libredirect -I nixpkgs=https://github.com/nixos/nixpkgs/archive/nixpkgs-21.11-darwin.tar.gz
/nix/store/19j8my6waz6agi4247jx3q23lx3wqx69-libredirect-0
```
trying to force a build on the above machine:
```
$ nix-build "<nixpkgs>" -A libredirect -I nixpkgs=https://github.com/nixos/nixpkgs/archive/nixpkgs-21.11-darwin.tar.gz --check
checking outputs of '/nix/store/n2jajggzh9rm02dkxdh2s65x5xsdycgn-libredirect-0.drv'...
unpacking sources
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
no configure script, doing nothing
building
libredirect.c:102:11: warning: incompatible function pointer types initializing 'int (*)(const char *, int, mode_t)' (aka 'int (*)(const char *, int, unsigned short)') with an expression of type 'int (*)(const char *, int, ...)' [-Wincompatible-function-pointer-types]
    int (*open_real) (const char *, int, mode_t) = LOOKUP_REAL(open);
          ^                                        ~~~~~~~~~~~~~~~~~
libredirect.c:107:27: warning: second argument to 'va_arg' is of promotable type 'mode_t' (aka 'unsigned short'); this va_arg has undefined behavior because arguments will be promoted to 'int' [-Wvarargs]
        mode = va_arg(ap, mode_t);
                          ^~~~~~
/nix/store/7sfdssd4p5c6gh5dr2j1fy06fxky195q-clang-wrapper-11.1.0/resource-root/include/stdarg.h:19:50: note: expanded from macro 'va_arg'
#define va_arg(ap, type)    __builtin_va_arg(ap, type)
                                                 ^~~~
libredirect.c:134:11: warning: incompatible function pointer types initializing 'int (*)(int, const char *, int, mode_t)' (aka 'int (*)(int, const char *, int, unsigned short)') with an expression of type 'int (*)(int, const char *, int, ...)' [-Wincompatible-function-pointer-types]
    int (*openat_real) (int, const char *, int, mode_t) = LOOKUP_REAL(openat);
          ^                                               ~~~~~~~~~~~~~~~~~~~
libredirect.c:139:27: warning: second argument to 'va_arg' is of promotable type 'mode_t' (aka 'unsigned short'); this va_arg has undefined behavior because arguments will be promoted to 'int' [-Wvarargs]
        mode = va_arg(ap, mode_t);
                          ^~~~~~
/nix/store/7sfdssd4p5c6gh5dr2j1fy06fxky195q-clang-wrapper-11.1.0/resource-root/include/stdarg.h:19:50: note: expanded from macro 'va_arg'
#define va_arg(ap, type)    __builtin_va_arg(ap, type)
                                                 ^~~~
4 warnings generated.
installing
install: creating directory '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0'
install: creating directory '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib'
'libredirect.dylib' -> '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib'
post-installation fixup
patching script interpreter paths in /nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0
patching script interpreter paths in /nix/store/cbzgblyik90wag4arznwrw5sx62xfqsa-libredirect-0-hook
running install tests
dyld[53552]: terminating because inserted dylib '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' could not be loaded: tried: '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/usr/local/lib/libredirect.dylib' (no such file), '/usr/lib/libredirect.dylib' (no such file)
dyld[53552]: tried: '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/usr/local/lib/libredirect.dylib' (no such file), '/usr/lib/libredirect.dylib' (no such file)
Assertion failed: (system(TESTPATH) == 0), function test_system, file test.c, line 37.
/nix/store/i97a6fwzdp2n2mb9crzhnvdb7r15arqi-stdenv-darwin/setup: line 1355: 53550 Abort trap: 6           NIX_REDIRECTS="/foo/bar/test=/nix/store/dwhdayp98xmpp9vzzb1qp4va1pjy65gz-coreutils-9.0/bin/true" ./test
error: builder for '/nix/store/n2jajggzh9rm02dkxdh2s65x5xsdycgn-libredirect-0.drv' failed with exit code 134;
       last 10 log lines:
       > install: creating directory '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib'
       > 'libredirect.dylib' -> '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib'
       > post-installation fixup
       > patching script interpreter paths in /nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0
       > patching script interpreter paths in /nix/store/cbzgblyik90wag4arznwrw5sx62xfqsa-libredirect-0-hook
       > running install tests
       > dyld[53552]: terminating because inserted dylib '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' could not be loaded: tried: '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/usr/local/lib/libredirect.dylib' (no such file), '/usr/lib/libredirect.dylib' (no such file)
       > dyld[53552]: tried: '/nix/store/27qmzv86dlmr9pibvmc6mb28n1f987qy-libredirect-0/lib/libredirect.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'arm64e')), '/usr/local/lib/libredirect.dylib' (no such file), '/usr/lib/libredirect.dylib' (no such file)
       > Assertion failed: (system(TESTPATH) == 0), function test_system, file test.c, line 37.
       > /nix/store/i97a6fwzdp2n2mb9crzhnvdb7r15arqi-stdenv-darwin/setup: line 1355: 53550 Abort trap: 6           NIX_REDIRECTS="/foo/bar/test=/nix/store/dwhdayp98xmpp9vzzb1qp4va1pjy65gz-coreutils-9.0/bin/true" ./test
       For full logs, run 'nix log /nix/store/n2jajggzh9rm02dkxdh2s65x5xsdycgn-libredirect-0.drv'.

--- a/pkgs/build-support/libredirect/default.nix
+++ b/pkgs/build-support/libredirect/default.nix
@@ -1,5 +1,4 @@
-{ stdenv, lib, coreutils }:
-
+{ lib, stdenv, bintools-unwrapped, llvmPackages_13, coreutils }:
 stdenv.mkDerivation rec {
   pname = "libredirect";
   version = "0";
@@ -9,17 +8,36 @@ stdenv.mkDerivation rec {
     cp ${./test.c} test.c
   '';
 
-  libName = "libredirect" + stdenv.targetPlatform.extensions.sharedLibrary;
-
   outputs = ["out" "hook"];
+
+  libName = "libredirect" + stdenv.targetPlatform.extensions.sharedLibrary;
 
   buildPhase = ''
     runHook preBuild
 
-    $CC -Wall -std=c99 -O3 -fPIC -ldl -shared \
-      ${lib.optionalString stdenv.isDarwin "-Wl,-install_name,$out/lib/$libName"} \
-      -o "$libName" \
-      libredirect.c
+    ${if stdenv.isDarwin && stdenv.isAarch64 then ''
+    # We need the unwrapped binutils and clang:
+    # We also want to build a fat library with x86_64, arm64, arm64e in there.
+    # Because we use the unwrapped tools, we need to provide -isystem for headers
+    # and the library search directory for libdl.
+    # We can't build this on x86_64, because the libSystem we point to doesn't
+    # like arm64(e).
+    PATH=${bintools-unwrapped}/bin:${llvmPackages_13.clang-unwrapped}/bin:$PATH \
+      clang -arch x86_64 -arch arm64 -arch arm64e \
+      -isystem ${llvmPackages_13.clang.libc}/include \
+      -isystem ${llvmPackages_13.libclang.lib}/lib/clang/*/include \
+      -L${llvmPackages_13.clang.libc}/lib \
+      -Wl,-install_name,$out/lib/$libName \
+      -Wall -std=c99 -O3 -fPIC libredirect.c \
+      -ldl -shared -o "$libName"
+    '' else if stdenv.isDarwin then ''
+    $CC -Wall -std=c99 -O3 -fPIC libredirect.c \
+      -Wl,-install_name,$out/lib/$libName \
+      -ldl -shared -o "$libName"
+    '' else ''
+    $CC -Wall -std=c99 -O3 -fPIC libredirect.c \
+      -ldl -shared -o "$libName"
+    ''}
 
     if [ -n "$doInstallCheck" ]; then
       $CC -Wall -std=c99 -O3 test.c -o test

